File: parser/tcp.py
from scapy.layers.inet import TCP
from scapy.packet import Packet as ScapyPacket

from parser.packet import Packet, TCPFlags


def parse_tcp(p: Packet, raw: ScapyPacket) -> None:
    if not raw.haslayer(TCP):
        return

    tcp = raw[TCP]
    p.protocol = "TCP"
    p.src_port = tcp.sport
    p.dst_port = tcp.dport
    p.payload = bytes(tcp.payload)

    flags = tcp.flags
    p.flags = TCPFlags(
        syn=bool(flags & 0x02),
        ack=bool(flags & 0x10),
        rst=bool(flags & 0x04),
        fin=bool(flags & 0x01),
        psh=bool(flags & 0x08),
        urg=bool(flags & 0x20),
    )

File: parser/tcp.go
package parser

import (
	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
)

func parseTCP(p *Packet, raw gopacket.Packet) {
	tcpLayer := raw.Layer(layers.LayerTypeTCP)
	if tcpLayer == nil {
		return
	}
	tcp := tcpLayer.(*layers.TCP)

	p.Protocol = "TCP"
	p.SrcPort = uint16(tcp.SrcPort)
	p.DstPort = uint16(tcp.DstPort)
	p.Payload = tcp.Payload
	p.Flags = TCPFlags{
		SYN: tcp.SYN,
		ACK: tcp.ACK,
		RST: tcp.RST,
		FIN: tcp.FIN,
		PSH: tcp.PSH,
		URG: tcp.URG,
	}
}
